src/legacy/ui/ui_settings/ui_settings_service.ts

```diff
@@ -65,82 +114,90 @@ export class UiSettingsService {
   }
 
   // returns a Promise for the value of the requested setting
-  async get(key) {
+  async get<T extends SavedObjectAttribute = any>(key: string): Promise<T> {
     const all = await this.getAll();
     return all[key];
   }
 
-  async getAll() {
+  async getAll<T extends SavedObjectAttribute = any>() {
     const raw = await this.getRaw();
 
-    return Object.keys(raw)
-      .reduce((all, key) => {
+    return Object.keys(raw).reduce(
+      (all, key) => {
         const item = raw[key];
-        const hasUserValue = 'userValue' in item;
-        all[key] = hasUserValue ? item.userValue : item.value;
+        all[key] = ('userValue' in item ? item.userValue : item.value) as T;
         return all;
-      }, {});
+      },
+      {} as Record<string, T>
+    );
   }
 
-  async getRaw() {
+  // NOTE: should be a private method
+  async getRaw(): Promise<UiSettingsRaw> {
     const userProvided = await this.getUserProvided();
     return defaultsDeep(userProvided, await this.getDefaults());
   }
 
-  async getUserProvided(options) {
-    const userProvided = {};
+  async getUserProvided(options: ReadOptions = {}): Promise<UserProvided> {
+    const userProvided: UserProvided = {};
 
     // write the userValue for each key stored in the saved object that is not overridden
     for (const [key, userValue] of Object.entries(await this._read(options))) {
       if (userValue !== null && !this.isOverridden(key)) {
         userProvided[key] = {
-          userValue
+          userValue,
         };
       }
     }
 
     // write all overridden keys, dropping the userValue is override is null and
     // adding keys for overrides that are not in saved object
     for (const [key, userValue] of Object.entries(this._overrides)) {
-      userProvided[key] = userValue === null
-        ? { isOverridden: true }
-        : { isOverridden: true, userValue };
+      userProvided[key] =
+        userValue === null ? { isOverridden: true } : { isOverridden: true, userValue };
     }
 
     return userProvided;
   }
 
-  async setMany(changes) {
+  async setMany<T extends SavedObjectAttribute = any>(changes: Record<string, T>) {
     await this._write({ changes });
   }
 
-  async set(key, value) {
+  async set<T extends SavedObjectAttribute = any>(key: string, value: T) {
```

Comment:
probably we should improve client side typings as well, right now they get/set `any` https://github.com/restrry/kibana/blob/33486f99d4c77cc9213f83cb4b7109ffa84207fe/src/core/public/ui_settings/ui_settings_client.ts#L125